### PR TITLE
Improve handling of indestructable blocks and headless pistons

### PR DESCRIPTION
**Pull request by commandz0**
Currently, all blocks being moved by a piston are indestructible by explosions unless the option `unsupported-settings.allow-permanent-block-break-exploits` is enabled (#4096). To prevent headless pistons from being created, this option also makes *all* moving piston blocks indestructible. With the current settings, it is not possible to allow moving blocks to be destroyed by TNT without also allowing bedrock to be broken.

This pull request splits the behavior of preventing headless pistons into its own config option (`unsupported-settings.allow-headless-pistons`). It also properly handles the destruction of a moving piston head by breaking the associated piston base, rather than making all moving blocks indestructible.

Finally, this pull request also adds other indestructible blocks to the list - command blocks, barriers, structure blocks, and jigsaw blocks. While these blocks do not naturally generate in a survival world, most admins would still wish to prevent them from being broken by survival players.

### CODE REVIEW

File: Spigot-Server-Patches/0552-Prevent-headless-pistons-from-being-created.patch [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+--- a/src/main/java/net/minecraft/server/Explosion.java
++++ b/src/main/java/net/minecraft/server/Explosion.java
+@@ -134,6 +134,15 @@ public class Explosion {
+ 
+                             if (f > 0.0F && this.k.a(this, this.world, blockposition, iblockdata, f) && blockposition.getY() < 256 && blockposition.getY() >= 0) { // CraftBukkit - don't wrap explosions
+                                 set.add(blockposition);
++                                // Paper start - prevent headless pistons from forming
++                                if (!com.destroystokyo.paper.PaperConfig.allowHeadlessPistons && iblockdata.getBlock() == Blocks.MOVING_PISTON) {
++                                    TileEntityPiston extension = (TileEntityPiston)this.world.getTileEntity(blockposition);
++                                    if (extension.h()) { // check if this is the piston head
```
Comment (aikar) [inline Spigot-Server-Patches/0552-Prevent-headless-pistons-from-being-created.patch]: needs an obfuscation helper for the .h()

**Review by aikar — DISMISSED** [review]
seems reasonable, just need the obf helper and wait for 1.16.2, but this can also be applied to 1.15.2 if you want to PR for that too.

**Comment by commandz0** [comment]
I've added the obfuscation helper.

How would I go about PRing to 1.15.2? Should I just repeat the same process for editing and creating patches, or is there some way to do it automatically through Git?

File: Spigot-Server-Patches/0552-Prevent-headless-pistons-from-being-created.patch [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                             d4 += d0 * 0.30000001192092896D;
+diff --git a/src/main/java/net/minecraft/server/TileEntityPiston.java b/src/main/java/net/minecraft/server/TileEntityPiston.java
+index e7b7e468fc874b0fbcd43f87d816a42420f9b05e..88ed2906f158837a1b7ea02f839ccf2baf468075 100644
+--- a/src/main/java/net/minecraft/server/TileEntityPiston.java
++++ b/src/main/java/net/minecraft/server/TileEntityPiston.java
+@@ -45,6 +45,8 @@ public class TileEntityPiston extends TileEntity implements ITickable {
+         return this.g;
+     }
+ 
++    public final boolean isHead() { return this.h(); } // Paper - OBFHELPER
```
Comment (Proximyst) [inline Spigot-Server-Patches/0552-Prevent-headless-pistons-from-being-created.patch]: Place this right above the `h` method.

**Review by Proximyst — DISMISSED** [review]
LGTM. Just need to move the obf helper to the method it's supposed to call.

**Review by Proximyst — DISMISSED** [review]

**Review by aurorasmiles — DISMISSED** [review]

**Review by Proximyst — DISMISSED** [review]

**Review by aurorasmiles — DISMISSED** [review]

**Review by kashike — DISMISSED** [review]

**Review by kashike — APPROVED** [review]

**Pull request merged**